lua/packer.lua

```diff
@@ -167,6 +168,108 @@ packer.use_rocks = function(rock)
   end
 end
 
+-- Merges new into old; returns merged spec.
+local function merge_specs(old, new)
+  local name = new.short_name
+
+  -- disable: Warn & ignore on disabling a previously enabled plugin
+  if old.disable then
+    if not new.disable then log.warn('Trying to disable previously non-disabled mod "' .. name .. '", ignoring') end
+  else
+    old.disable = new.disable
+  end
+
+  -- branch, tag, commit, lock: Must all be the same, else warning
+  for _, key in ipairs{'branch', 'commit', 'lock', 'tag'} do
+    if new[key] ~= old[key] then
+      log.warn('Plugin "' .. name .. '" wants to overwrite "' .. key .. '" setting. Keeping original value')
+    end
+  end
+
+  -- opt
+  -- - Conditions and opt_default have already been handled in manage()
+  -- - Rule of thumb: true < nil < false. Set the lowest necessary--if nil suffices, don't set false
+  -- - For manual_opt:
+  --   - If result is start, remove mo. Else, if omo or nmo, keep mo.
+  --   - There is no situation where opt = falsy but manual_opt = true
+  --
+  -- old new result (1 = true, - = nil, 0 = false, s = start, o = opt)
+  -- ---+---+------
+  --   -   -     -s
+  --   1   -     -s
+  --   -   1     -s
+  --   0   0     0s
+  --   1   0     0s
+  --   0   1     0s
+  --   -   0     0s
+  --   0   -     0s
+  --   1   1     1o
+  local old_opt = old.opt
+  local new_opt = new.opt
+  if old_opt == false or new_opt == false then
+    old.opt = false
+    old.manual_opt = nil
+  elseif old_opt == nil or new_opt == nil then
+    old.opt = nil
+    old.manual_opt = nil
+  elseif old_opt == true and new_opt == true then
+    if old.manual_opt or new.manual_opt then old.manual_opt = true end
+  end
+
+  -- config, setup: append
+  for _, key in ipairs{'config', 'setup'} do
+    if new[key] then
+      if type(old[key]) ~= 'table' then old[key] = {old[key]} end
+      if type(new[key]) ~= 'table' then new[key] = {new[key]} end
+      for _, item in ipairs(new[key]) do
+        table.insert(old[key], item)
+      end
+    end
+  end
+
+  -- installer, updater: Warn & ignore
+  for _, key in ipairs{'installer', 'updater'} do
+    if new[key] then
+      log.warn('Plugin "' .. name .. '" wants to overwrite "' .. key .. '" setting. Keeping original value')
+    end
+  end
+
+  -- run, rtp: Keep old key if it's set, otherwise set new one
+  for _, key in ipairs{'run', 'rtp'} do
+    if new[key] then
+      if old[key] then
+        local old_type = type(old[key])
+        local new_type = type(new[key])
+        if (old_type ~= new_type) or
+            -- If the first condition doesn't hit, old and new have the same type
+            (old_type == 'string' and old[key] ~= new[key]) or
+            (old_type == 'function' and (string.dump(old[key], true) ~= string.dump(new[key], true)))
```

Comment:
Does direct function comparison not work here? I suppose users are most likely to specify anonymous functions which will compare unequal, but unless the user has specified exactly the same function in the new spec as the old, the bytecode will compare unequal too.

Reply:
True, I'll change it.